DAOS-5798 dfs: enhance support for reading symlinks

daltonbohning opened the pull request:
- Moved core logic of dfs_lookup into lookup_rel_path
- lookup_rel_path conditionally dereferences leaf symbolic
links and implements partial support for ".."
  - Recursion up to DFS_MAX_RECURSION = 40
- Changed dfuse calls to dfs_lookup_rel to pass O_NOFOLLOW
- Added dfs unit tests for dfs_lookup, dfs_lookup_rel, and the O_NOFOLLOW behavior.

Signed-off-by: Dalton Bohning <daltonx.bohning@intel.com>

daltonbohning commented:
@mchaarawi @ashleypittman I'm working on this PR in conjunction with corresponding changes in mpifileutils. I'd appreciate any feedback on the current changes to make sure the changes to mpifileutils are valid.
For now, I have a new functional test introduced in this PR for convenience, but that can be separated into a separate PR if needed and merged after this PR.

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1785,9 +1786,9 @@ dfs_remove(dfs_t *dfs, dfs_obj_t *parent, const char *name, bool force,
-dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-	   mode_t *mode, struct stat *stbuf)
+static int
+lookup_rel_path(dfs_t *dfs, dfs_obj_t *root, const char *path, int flags, dfs_obj_t **_obj,
```
(style)  line over 80 characters

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1798,23 +1799,11 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-	}
+	/* Only paths from root can be absolute */
+	if (path[0] == '/' && daos_oid_cmp(root->oid, dfs->root.oid) != 0)
+ 		return EINVAL;
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
		return EINVAL;
```

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1849,7 +1838,47 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-dfs_lookup_loop:
+lookup_rel_path_loop:
+
+		/* 
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion
		/*
```

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1916,27 +1945,61 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+					D_FREE(entry.value);
+					D_GOTO(err_obj, rc);
+				}
+				
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion

```

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1997,6 +2060,28 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
 
+int
+dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+           mode_t *mode, struct stat *stbuf)
```
(style)  code indent should use tabs where possible

daosbuild1 replied:
```suggestion
	   mode_t *mode, struct stat *stbuf)
```

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -3101,7 +3203,7 @@ dfs_chmod(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode)
 		D_ASSERT(entry.value);
 
-		rc = dfs_lookup(dfs, entry.value, O_RDWR, &sym, NULL, NULL);
+		rc = lookup_rel_path(dfs, parent, entry.value, O_RDWR, &sym, NULL, NULL);
```
(style)  line over 80 characters

daosbuild1 commented on src/tests/ftest/datamover/copy_symlinks.py:
```diff
@@ -0,0 +1,300 @@
+        cmd = "; \\\n".join(cmd_list)
+        self.execute_cmd(cmd)
+
+    # TODO move this to TestBase and formalize
```
(pylint-fixme) TODO move this to TestBase and formalize

daosbuild1 commented on src/tests/ftest/datamover/copy_symlinks.py:
```diff
@@ -0,0 +1,300 @@
+  portions thereof marked with this legend must also reproduce the markings.
+'''
+from data_mover_test_base import DataMoverTestBase
+from os.path import join, sep
```
(pylint-unused-import) Unused sep imported from os.path

daosbuild1 commented on src/tests/ftest/util/data_mover_test_base.py:
```diff
@@ -177,6 +188,16 @@ def gen_daos_path(self, prefix=None):
             return join(prefix, daos_dir)
         return join("/", daos_dir)
 
+    def new_posix_dir(self):
```
(pylint-missing-docstring) Missing method docstring

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3970/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/1/execution/node/57/log

daosbuild1 commented on src/tests/ftest/datamover/copy_symlinks.py:
```diff
@@ -0,0 +1,300 @@
+        cmd = "; \\\n".join(cmd_list)
+        self.execute_cmd(cmd)
+
+    # TODO move this to TestBase and formalize
```
(pylint-fixme) TODO move this to TestBase and formalize

daosbuild1 replied:
(pylint-fixme) TODO move this to TestBase and formalize

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3970/2/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/2/execution/node/58/log

daosbuild1 commented:
Test stage Build on Ubuntu 20.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/2/execution/node/307/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/2/execution/node/313/log

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3970/3/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/3/execution/node/58/log

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/3/execution/node/1025/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi commented on src/client/dfs/dfs.c:
```diff
@@ -1997,6 +2059,28 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+	   mode_t *mode, struct stat *stbuf)
+{
+	if (dfs == NULL || !dfs->mounted)
+	return EINVAL;
```
missing a tab?

mchaarawi commented on src/client/dfs/dfs.c:
```diff
@@ -1849,7 +1838,47 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+		 * nor can we support "../.." because we don't
+		 * have parent.parent_oid and parent.mode.
+		 * For now, represent this partial state with
+		 * parent_fully_valid.
```
i have some thought on how to support .., but unfortunately it required a layout change and some modifications to rename. but something for the future

daltonbohning replied:
Sure. It's an edge case anyway

mchaarawi commented on src/client/dfs/dfs.c:
```diff
@@ -1849,7 +1838,47 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+
+			/* Cannot go outside the container */
+			if (daos_oid_cmp(parent.oid, dfs->root.oid) == 0) {
+				D_GOTO(err_obj, rc = ENOENT);
```
not required, but printing a debug message here might be helpful

daltonbohning replied:
I agree. I added one

mchaarawi commented:
im not sure if it's doing that, but all functional tests should run with this PR

daltonbohning commented:
> im not sure if it's doing that, but all functional tests should run with this PR before you request landing

Right. I can't run the new functional tests yet, since the changes to mpifileutils have yet to be merged upstream, and thus also not in CI, so they would fail.

The goal is to:
- Get the DFS changes approved
- Merge the mpifileutils changes upstream
- Update mpifileutils in CI
- Land this PR

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented:
I realized I completely forgot to check for infinite recursion. Added that, as well as some memory fixes.

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1805,9 +1806,9 @@ dfs_remove(dfs_t *dfs, dfs_obj_t *parent, const char *name, bool force,
-	   mode_t *mode, struct stat *stbuf)
+static int
+lookup_rel_path(dfs_t *dfs, dfs_obj_t *root, const char *path, int flags,
+		dfs_obj_t **_obj, mode_t *mode, struct stat *stbuf, size_t depth)
```
(style)  line over 80 characters

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1936,27 +1972,63 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+				}
+
+				rc = lookup_rel_path(dfs, &parent, entry.value,
+						     flags, &sym, NULL, NULL, depth + 1);
```
(style)  line over 80 characters

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -1936,27 +1972,63 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+				}
+
+				rc = lookup_rel_path(dfs, &parent, entry.value,
+						     flags, &sym, mode, stbuf, depth + 1);
```
(style)  line over 80 characters

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -2017,6 +2089,28 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+		path += dfs->prefix_len;
+	}
+
+	return lookup_rel_path(dfs, &dfs->root, path, flags, _obj, mode, stbuf, 0);
```
(style)  line over 80 characters

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3970/7/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/7/execution/node/60/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented:
@mchaarawi 
1. I updated lookup_rel_path with a "depth" argument, to prevent infinite recursion. I put this as the last argument, even though it's only input, not output. My reasoning was that it's really more of a "meta" argument. Let me know if this isn't in line with our style.
2. I removed the functional tests due to some dependencies that aren't yet ready. They will be introduced separately.
3. I added unit tests to verify the new changes, as well as dfs_lookup[_rel] in general.

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -123,6 +245,113 @@ dfs_test_syml(void **state)
+	strncpy(path, "/", 2);
+	strncat(path, name, 62);
+
+	// O_NOFOLLOW should open the link itself
```
(style)  do not use C99 // comments

daosbuild1 replied:
```suggestion
	/* O_NOFOLLOW should open the link itself */
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -123,6 +245,113 @@ dfs_test_syml(void **state)
+	assert_int_equal(rc, 0);
+	assert_true(S_ISLNK(actual_mode));
+
+	// O_NOFOLLOW should open the link itself
```
(style)  do not use C99 // comments

daosbuild1 replied:
```suggestion
	/* O_NOFOLLOW should open the link itself */
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -123,6 +245,113 @@ dfs_test_syml(void **state)
+	assert_int_equal(rc, 0);
+	assert_true(S_ISLNK(actual_mode));
+
+	// Default should follow the link
```
(style)  do not use C99 // comments

daosbuild1 replied:
```suggestion
	/* Default should follow the link */
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -123,6 +245,113 @@ dfs_test_syml(void **state)
+	assert_int_equal(rc, 0);
+	assert_int_equal(actual_mode & S_IFMT, mode & S_IFMT);
+
+	// Default should follow the link
```
(style)  do not use C99 // comments

daosbuild1 replied:
```suggestion
	/* Default should follow the link */
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -84,6 +84,128 @@ dfs_test_mount(void **state)
+
+	/** Create /sym1 -> dir1 */
+	rc = dfs_open(dfs_mt, NULL, filename_sym1, create_mode | S_IFLNK,
+			create_flags, 0, 0, filename_dir1, &obj);
```
```suggestion
		      create_flags, 0, 0, filename_dir1, &obj);
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -84,6 +84,128 @@ dfs_test_mount(void **state)
+
+	/** Create /dir1 */
+	rc = dfs_open(dfs_mt, NULL, filename_dir1, create_mode | S_IFDIR,
+		create_flags, 0, 0, NULL, &dir);
```
```suggestion
		      create_flags, 0, 0, NULL, &dir);
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -84,6 +84,128 @@ dfs_test_mount(void **state)
+
+	/** Create /dir1/dir2 */
+	rc = dfs_open(dfs_mt, dir, filename_dir2, create_mode | S_IFDIR,
+			create_flags, 0, 0, NULL, &obj);
```
```suggestion
		      create_flags, 0, 0, NULL, &obj);
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -84,6 +84,128 @@ dfs_test_mount(void **state)
+
+	/** Create /dir1/file2 */
+	rc = dfs_open(dfs_mt, dir, filename_file2, create_mode | S_IFREG,
+			create_flags, 0, 0, NULL, &obj);
```
```suggestion
		      create_flags, 0, 0, NULL, &obj);
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -123,6 +245,113 @@ dfs_test_syml(void **state)
+
+	/** Create /dir */
+	rc = dfs_open(dfs_mt, NULL, name_dir, S_IFDIR | S_IWUSR | S_IRUSR,
+			O_RDWR | O_CREAT | O_EXCL, 0, 0, NULL, &obj);
```
```suggestion
		      O_RDWR | O_CREAT | O_EXCL, 0, 0, NULL, &obj);
```

daosbuild1 commented on src/tests/suite/dfs_unit_test.c:
```diff
@@ -123,6 +245,113 @@ dfs_test_syml(void **state)
+
+	/** Create /sym_to_dir -> dir */
+	rc = dfs_open(dfs_mt, NULL, name_sym_to_dir, S_IFLNK,
+			O_RDWR | O_CREAT | O_EXCL, 0, 0, name_dir, &obj);
```
```suggestion
		      O_RDWR | O_CREAT | O_EXCL, 0, 0, name_dir, &obj);
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3970/10/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/10/execution/node/61/log

daltonbohning commented on src/client/dfs/dfs.c:
```diff
@@ -3024,7 +3144,8 @@ dfs_access(dfs_t *dfs, dfs_obj_t *parent, const char *name, int mask)
 	D_ASSERT(entry.value);
 
-	rc = dfs_lookup(dfs, entry.value, O_RDONLY, &sym, NULL, NULL);
+	rc = lookup_rel_path(dfs, parent, entry.value, O_RDONLY, &sym,
```
Ah, yes, it does. I just added another call in the unit tests to verify this as well.

daltonbohning commented on src/client/dfs/dfs.c:
```diff
@@ -1818,23 +1820,15 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-	if (path[0] != '/')
-		return EINVAL;
+	/* Arbitrarily stop recursion after 40 levels */
+	if (depth >= 40)
```
Agreed. Added, and a unit test for ELOOP.

daltonbohning commented on src/client/dfs/dfs.c:
```diff
@@ -1818,23 +1820,15 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-		path += dfs->prefix_len;
-	}
+	/* Only paths from root can be absolute */
+	if (path[0] == '/' && daos_oid_cmp(root->oid, dfs->root.oid) != 0)
```
Yes, it is necessary. When `dfs_lookup` is called, the '/' is ALWAYS required (previous behavior).
But `lookup_rel_path` can be called recursively with a _**relative**_ path that does not start with '/', but for which the parent is the root. I couldn't think of a cleaner way than to have the duplicate check, other than to check right before making a recursive call, which would require a check at each place that calls `lookup_rel_path`.
I'm open to other suggestions, if you have any.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented:
Updated per comments and rebuilt against newer master.
@ashleypittman This doesn't seem to have broken anything in DFUSE, but please let me know if you see anything fishy.

daltonbohning commented on src/client/dfs/dfs.c:
```diff
@@ -1805,9 +1809,10 @@ dfs_remove(dfs_t *dfs, dfs_obj_t *parent, const char *name, bool force,
-int
-dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-	   mode_t *mode, struct stat *stbuf)
+static int
+lookup_rel_path(dfs_t *dfs, dfs_obj_t *root, const char *path, int flags,
+		dfs_obj_t **_obj, mode_t *mode, struct stat *stbuf,
+		size_t depth)
```
@ashleypittman This is the function I mentioned before. In order to handle symlink dereferencing, I needed to be able to lookup a path relative to some parent, which is what this function does. It's literally just `dfs_lookup`, but relative to some `root`.

mchaarawi commented:
you need a merge with master

daltonbohning commented:
> you need a merge with master

@mchaarawi Just pushed. Thanks!

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/13/execution/node/706/log

daosbuild1 commented:
Test stage NLT completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/14/execution/node/341/log

ashleypittman commented:
The NLT error is reported at https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-3970/14/valgrindResult/pid=39227,0x18/

ashleypittman commented on src/client/dfs/dfs.c:
```diff
@@ -2255,15 +2358,33 @@ dfs_lookup_rel(dfs_t *dfs, dfs_obj_t *parent, const char *name, int flags,
-			stbuf->st_size = entry.value_len;
+		if (flags & O_NOFOLLOW) {
+			/* Create a truncated version of the string */
+			D_STRNDUP(obj->value, entry.value, entry.value_len + 1);
```
Seems the segfault is because entry.value is NULL at this point.

daltonbohning replied:
Thanks @ashleypittman . It looks like I messed up the merge here. There should be a break statement before the S_IFLNK case. The S_IFREG is falling through here

ashleypittman replied:
I saw you'd retriggerd the Unit Test phase and I didn't think that was going to work.

daltonbohning replied:
For some reason, when I looked at the logs, I was seeing some erroneous error, so I just restarted. But I think maybe I had just clicked the wrong link :)

daosbuild1 commented on src/client/dfs/dfs.c:
```diff
@@ -2255,15 +2358,34 @@ dfs_lookup_rel(dfs_t *dfs, dfs_obj_t *parent, const char *name, int flags,
 			stbuf->st_blocks = (stbuf->st_size + (1 << 9) - 1) >> 9;
 		}
-		break;
+        break;
```
(style)  code indent should use tabs where possible

daosbuild1 replied:
```suggestion
	break;
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3970/15/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/15/execution/node/60/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on CentOS 7 release completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/16/execution/node/349/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3970/16/execution/node/292/log

daltonbohning commented:
@ashleypittman @mchaarawi Finally got this through CI... The only changes are really just the merge with master.
